File: app/src/main/java/com/jasonwiram/golfscorecard/ListAdapter.java
package com.jasonwiram.golfscorecard;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.TextView;

public class ListAdapter extends BaseAdapter {
    private final Context mContext;
    private final Hole[] mHoles;

    public ListAdapter (Context context, Hole[] holes) {
        mContext = context;
        mHoles = holes;
    }

    @Override
    public int getCount() {
        return mHoles.length;
    }

    @Override
    public Object getItem(int i) {
        return mHoles[i];
    }

    @Override
    public long getItemId(int i) {
        return 0;
    }

    @Override
    public View getView(final int i, View view, ViewGroup viewGroup) {
        final ViewHolder holder;

        if (view == null) {
            view = LayoutInflater.from(mContext).inflate(R.layout.list_item, null);

            holder = new ViewHolder();
            holder.holeLabel = (TextView) view.findViewById(R.id.holeLabel);
            holder.strokeCount = (TextView) view.findViewById(R.id.strokeCount);
            holder.removeStrokeButton = (Button) view.findViewById(R.id.removeStrokeButton);
            holder.addStrokeButton = (Button) view.findViewById(R.id.addStrokeButton);

            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }

        holder.holeLabel.setText(mHoles[i].getmLabel());
        holder.strokeCount.setText(mHoles[i].getmStrokeCount() + "");
        holder.removeStrokeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int updatedStrokeCount = mHoles[i].getmStrokeCount() - 1;
                if (updatedStrokeCount < 0) updatedStrokeCount = 0;
                mHoles[i].setmStrokeCount(updatedStrokeCount);
                holder.strokeCount.setText(updatedStrokeCount + "");
            }
        });
        holder.addStrokeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int updatedStrokeCount = mHoles[i].getmStrokeCount() + 1;
                mHoles[i].setmStrokeCount(updatedStrokeCount);
                holder.strokeCount.setText(updatedStrokeCount + "");
            }
        });

        return view;
    }

    private static class ViewHolder {
        TextView holeLabel;
        TextView strokeCount;
        Button removeStrokeButton;
        Button addStrokeButton;
    }
}

File: app/src/main/java/com/jasonwiram/golfscorecard/ListAdapter.kt
package com.jasonwiram.golfscorecard

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.TextView

class ListAdapter(private val mContext: Context, private val mHoles: Array<Hole?>) : BaseAdapter() {

    override fun getCount(): Int {
        return mHoles.size
    }

    override fun getItem(i: Int): Hole? {
        return mHoles[i]
    }

    override fun getItemId(i: Int): Long {
        return 0
    }

    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
        var view = view
        val holder: ViewHolder

        if (view == null) {
            view = LayoutInflater.from(mContext).inflate(R.layout.list_item, null)

            holder = ViewHolder()
            holder.holeLabel = view!!.findViewById(R.id.holeLabel) as TextView
            holder.strokeCount = view.findViewById(R.id.strokeCount) as TextView
            holder.removeStrokeButton = view.findViewById(R.id.removeStrokeButton) as Button
            holder.addStrokeButton = view.findViewById(R.id.addStrokeButton) as Button

            view.tag = holder
        } else {
            holder = view.tag as ViewHolder
        }

        holder.holeLabel!!.text = mHoles[i]!!.getmLabel()
        holder.strokeCount!!.text = mHoles[i]!!.getmStrokeCount().toString() + ""
        holder.removeStrokeButton!!.setOnClickListener {
            var updatedStrokeCount = mHoles[i]!!.getmStrokeCount() - 1
            if (updatedStrokeCount < 0) updatedStrokeCount = 0
            mHoles[i]!!.setmStrokeCount(updatedStrokeCount)
            holder.strokeCount!!.text = updatedStrokeCount.toString() + ""
        }
        holder.addStrokeButton!!.setOnClickListener {
            val updatedStrokeCount = mHoles[i]!!.getmStrokeCount() + 1
            mHoles[i]!!.setmStrokeCount(updatedStrokeCount)
            holder.strokeCount!!.text = updatedStrokeCount.toString() + ""
        }

        return view
    }

    private class ViewHolder {
        internal var holeLabel: TextView? = null
        internal var strokeCount: TextView? = null
        internal var removeStrokeButton: Button? = null
        internal var addStrokeButton: Button? = null
    }
}
